### PR TITLE
Add `start`, `end` and `word_start` matching startegy

### PR DESCRIPTION
**Pull request by ahlee2326**
#287

### CODE REVIEW

**Comment by dunglas** [comment]
Run PHP CS Fixer to make Travis happy.

File: Doctrine/Orm/Filter/SearchFilter.php [hunk 1]
```diff
@@ -15,6 +15,7 @@
 use Doctrine\ORM\QueryBuilder;
 use Dunglas\ApiBundle\Api\IriConverterInterface;
 use Dunglas\ApiBundle\Api\ResourceInterface;
+use InvalidArgumentException;
```
Comment (sroze) [inline Doctrine/Orm/Filter/SearchFilter.php]: The FQCN is `Dunglas\ApiBundle\Exception\InvalidArgumentException` to use our "domain exception".

File: Doctrine/Orm/Filter/SearchFilter.php [hunk 2]
```diff
@@ -33,6 +34,14 @@ class SearchFilter extends AbstractFilter
      * @var string The value must be contained in the field.
      */
     const STRATEGY_PARTIAL = 'partial';
+    /**
+     * @var string Finds fields that are starting with the specified value.
+     */
+    const STRATEGY_STARTS_WITH = 'startsWith';
```
Comment (sroze) [inline Doctrine/Orm/Filter/SearchFilter.php]: I agree with @dunglas that snake_case is more coherent.

**Comment by sroze** [comment]
Thank you for your contribution @ahlee2326, not so far from merging :)

To summarise, it's missing:
- [ ] The code style fixes
- [ ] The exception namespace
- [ ] The smake cased query parameter.
- [ ] The Behat tests

**Comment by ahlee2326** [comment]
@sroze 
1. Fixer loves it
2. Fixed
3. Updated
4. Behat loves it

**Comment by sroze** [comment]
@ahlee2326 good work thank you! LGTM :+1: 

By the way, it would be great if the API user could specify that strategy in the request too.

**Comment by teohhanhui** [comment]
What about a `wordStarsWith` strategy for matching e.g. `Verge` in `The Verge`?

File: Doctrine/Orm/Filter/SearchFilter.php [hunk 3]
```diff
@@ -33,6 +34,14 @@ class SearchFilter extends AbstractFilter
      * @var string The value must be contained in the field.
      */
     const STRATEGY_PARTIAL = 'partial';
+    /**
+     * @var string Finds fields that are starting with the specified value.
+     */
+    const STRATEGY_START_WITH = 'start_with';
```
Comment (teohhanhui) [inline Doctrine/Orm/Filter/SearchFilter.php]: I think this should be `starts_with`.

File: Doctrine/Orm/Filter/SearchFilter.php [hunk 4]
```diff
@@ -33,6 +34,14 @@ class SearchFilter extends AbstractFilter
...
+    /**
+     * @var string Finds fields that are ending with the specified value.
+     */
+    const STRATEGY_END_WITH = 'end_with';
```
Comment (teohhanhui) [inline Doctrine/Orm/Filter/SearchFilter.php]: And `ends_with`.

File: Doctrine/Orm/Filter/SearchFilter.php (hunk 3 above)
Comment (teohhanhui) [inline Doctrine/Orm/Filter/SearchFilter.php]: Or do we go with just `start`?

File: Doctrine/Orm/Filter/SearchFilter.php (hunk 4 above)
Comment (dunglas) [inline Doctrine/Orm/Filter/SearchFilter.php]: the shorter the better

Comment (sroze) [inline Doctrine/Orm/Filter/SearchFilter.php]: :+1:, `end` is enough.

**Comment by sroze** [comment]
@ahlee2326 can you just replace `start_with` by `start` and `end_with` by `end` ?

**Comment by dunglas** [comment]
ping @ahlee2326

**Comment by ahlee2326** [comment]
@dunglas updated.

**Comment by dunglas** [comment]
Thank you @ahlee2326. Can you also update the doc at https://github.com/api-platform/doc?

**Pull request merged**